Kernel 5.2 fails to build with broadcom_sta

**Describe the bug**

Trying to build with `boot.kernelPackages = pkgs.linuxPackages_5_2;` and

```
boot.kernelModules = [ "kvm-amd" "wl" ];
boot.extraModulePackages = [ config.boot.kernelPackages.broadcom_sta ];
```

Results in build failure.

**To Reproduce**
Steps to reproduce the behavior:

Include the aforementioned options in configuration.nix and rebuild.

**Expected behavior**
Build completes successfully.

**Actual behavior**

Build fails:

```
(more similar warnings up here)
In file included from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/huge_mm.h:8:0,
                 from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/mm.h:570,
                 from /build/broadcom-sta/src/include/linuxver.h:65,
                 from /build/broadcom-sta/src/shared/linux_osl.c:25:
/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/fs.h:2959:16: note: expected 'size_t {aka long unsigned int}' but argument is of type 'char *'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^~~~~~~~~~~
/build/broadcom-sta/src/shared/linux_osl.c:1083:42: warning: passing argument 4 of 'kernel_read' makes pointer from integer without a cast [-Wint-conversion]
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                                          ^~~
In file included from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/huge_mm.h:8:0,
                 from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/mm.h:570,
                 from /build/broadcom-sta/src/include/linuxver.h:65,
                 from /build/broadcom-sta/src/shared/linux_osl.c:25:
/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/fs.h:2959:16: note: expected 'loff_t * {aka long long int *}' but argument is of type 'int'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^~~~~~~~~~~
  CC [M]  /build/broadcom-sta/src/wl/sys/wl_linux.o
/build/broadcom-sta/src/wl/sys/wl_linux.c: In function 'wl_pci_probe':
/build/broadcom-sta/src/wl/sys/wl_linux.c:781:2: warning: this 'if' clause does not guard... [-Wmisleading-indentation]
  if ((val & 0x0000ff00) != 0)
  ^~
/build/broadcom-sta/src/wl/sys/wl_linux.c:783:3: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'if'
   bar1_size = pci_resource_len(pdev, 2);
   ^~~~~~~~~
  CC [M]  /build/broadcom-sta/src/wl/sys/wl_iw.o
  CC [M]  /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.o
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c: In function 'wl_dev_ioctl':
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:460:9: error: implicit declaration of function 'get_ds'; did you mean 'get_fs'? [-Werror=implicit-function-declaration]
  set_fs(get_ds());
         ^~~~~~
         get_fs
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:460:9: error: incompatible type for argument 1 of 'set_fs'
In file included from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/uaccess.h:11:0,
                 from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/arch/x86/include/asm/fpu/xstate.h:5,
                 from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/arch/x86/include/asm/pgtable.h:26,
                 from /nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/include/linux/mm.h:99,
                 from /build/broadcom-sta/src/include/linuxver.h:65,
                 from /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:26:
/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/arch/x86/include/asm/uaccess.h:29:20: note: expected 'mm_segment_t {aka struct <anonymous>}' but argument is of type 'int'
 static inline void set_fs(mm_segment_t fs)
                    ^~~~~~
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c: In function 'wl_bss_roaming_done':
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:2378:2: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  struct cfg80211_roam_info roam_info = {
  ^~~~~~
cc1: some warnings being treated as errors
make[3]: *** [/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/scripts/Makefile.build:279: /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.o] Error 1
make[2]: *** [/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/Makefile:1597: _module_/build/broadcom-sta] Error 2
make[1]: *** [/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/source/Makefile:179: sub-make] Error 2
make[1]: Leaving directory '/nix/store/v6fb7f7s3mgv8ckywmskblqvjc8al95i-linux-5.2.14-dev/lib/modules/5.2.14/build'
make: *** [Makefile:161: all] Error 2
builder for '/nix/store/vjjww9mp7ch9hyk9s7qrymrld27acya9-broadcom-sta-6.30.223.271-5.2.14.drv' failed with exit code 2
cannot build derivation '/nix/store/z6bbhry3435ac08svlxvf3aicxrk3z78-kernel-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/s1vaiwsmvxa8w20alz1aswkvah267scl-nixos-system-MaFAS-0.15.9-19.03.173486.cf018a7c558.drv': 1 dependencies couldn't be built
error: build of '/nix/store/s1vaiwsmvxa8w20alz1aswkvah267scl-nixos-system-*hostname*-0.15.9-19.03.173486.cf018a7c558.drv' failed
```

From what I can tell, building with 5.2 includes stricter compiler options that broadcom_sta doesn't comply with.

**Metadata**

- system: `"x86_64-linux"`
 - host os: `Linux 4.19.72, NixOS, 19.03.173486.cf018a7c558 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(werner): `"nixos-18.09.2574.a7e559a5504"`
 - channels(root): `"nixos-19.03.173486.cf018a7c558"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`


Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
  - boot.kernelPackages
  - boot.kernelModules
  - boot.extraModulePackages
```

--- a/pkgs/os-specific/linux/broadcom-sta/default.nix
+++ b/pkgs/os-specific/linux/broadcom-sta/default.nix
@@ -34,6 +34,7 @@ stdenv.mkDerivation {
     # source: https://aur.archlinux.org/cgit/aur.git/tree/linux412.patch?h=broadcom-wl
     ./linux-4.12.patch
     ./linux-4.15.patch
+    ./linux-5.1.patch
     ./null-pointer-fix.patch
     ./gcc.patch
   ];

--- a/pkgs/os-specific/linux/broadcom-sta/linux-5.1.patch
+++ b/pkgs/os-specific/linux/broadcom-sta/linux-5.1.patch
@@ -1,0 +1,32 @@
+commit bcb06af629a36eb84f9a35ac599ec7e51e2d39fb
+Author: georgewhewell <georgerw@gmail.com>
+Date:   Sat May 18 21:22:37 2019 +0100
+
+    find src -type f -name \'*.c\' -exec sed -i "s/get_ds()/KERNEL_DS/g" {} \;
+
+diff --git a/src/wl/sys/wl_cfg80211_hybrid.c b/src/wl/sys/wl_cfg80211_hybrid.c
+index 7b606e0..51c81bc 100644
+--- a/src/wl/sys/wl_cfg80211_hybrid.c
++++ b/src/wl/sys/wl_cfg80211_hybrid.c
+@@ -450,7 +450,7 @@ wl_dev_ioctl(struct net_device *dev, u32 cmd, void *arg, u32 len)
+ 	ifr.ifr_data = (caddr_t)&ioc;
+ 
+ 	fs = get_fs();
+-	set_fs(get_ds());
++	set_fs(KERNEL_DS);
+ #if defined(WL_USE_NETDEV_OPS)
+ 	err = dev->netdev_ops->ndo_do_ioctl(dev, &ifr, SIOCDEVPRIVATE);
+ #else
+diff --git a/src/wl/sys/wl_iw.c b/src/wl/sys/wl_iw.c
+index c4c610b..9c3c74e 100644
+--- a/src/wl/sys/wl_iw.c
++++ b/src/wl/sys/wl_iw.c
+@@ -117,7 +117,7 @@ dev_wlc_ioctl(
+ 	ifr.ifr_data = (caddr_t) &ioc;
+ 
+ 	fs = get_fs();
+-	set_fs(get_ds());
++	set_fs(KERNEL_DS);
+ #if defined(WL_USE_NETDEV_OPS)
+ 	ret = dev->netdev_ops->ndo_do_ioctl(dev, &ifr, SIOCDEVPRIVATE);
+ #else

